Оплата через DMS

NadirCianid opened the pull request.

NadirCianid committed:
Создал миграцию и сущность для платежной операции

NadirCianid committed:
Добавил репозиторий для PaymentOperation

NadirCianid committed:
Переименовал сущность PaymentOperation в PaymentRequest

NadirCianid committed:
Начал писать метод для отправки авторизационного платежа

NadirCianid committed:
Реализовал отправку платежа (authorization message)

NadirCianid committed:
Реализовал отмену и подтверждение платежа (clearing message)

RickHammerson commented on src/main/java/faang/school/paymentservice/PaymentApplication.java:
```diff
@@ -1,11 +1,22 @@
+
+    @Bean
+    public ObjectMapper objectMapper() {
+        return new ObjectMapper();
```
То же замечание с другого PR
Чекни, может, я ошибаюсь

RickHammerson commented on src/main/java/faang/school/paymentservice/config/redis/RedisConfig.java:
```diff
@@ -0,0 +1,43 @@
+public class RedisConfig {
+    private int port;
+    private String host;
+    private String channel;
```
Всё таки, я бы выделил отдельно конфиг Redisа, где создаются топики и бины для Redisа самого и его Property
А то всё в кашу

Не забудь чекнуть, что написал про Channelы в другом PR

RickHammerson commented on src/main/java/faang/school/paymentservice/controller/PaymentController.java:
```diff
@@ -1,37 +1,40 @@
-        String message = String.format("Dear friend! Thank you for your purchase! " +
-                        "Your payment on %s %s was accepted.",
-                formattedSum, dto.currency().name());
+    public PaymentResponse sendPayment(@RequestBody @Validated PaymentRequestDto dto) {
```
Разве не @Valid?

RickHammerson commented on src/main/java/faang/school/paymentservice/controller/PaymentController.java:
```diff
@@ -1,37 +1,40 @@
-                message)
-        );
+    @PatchMapping("/payment/{paymentId}/confirm")
+    public PaymentResponse confirmPayment(@PathVariable @NotNull Long paymentId, @RequestParam(required = false) BigDecimal newAmount) {
```
Не уверен, что сработает аннотация @NotNull
Чекни плиз
Если не сработает, попробуй на класс навесить @Validated

RickHammerson replied:
И ещё интересно, почему там Required = false?

NadirCianid replied:
> И ещё интересно, почему там Required = false?

Клиент обычно не меняет сумму платежа. Но я решил, что такую возможность дать можно. Такое иногда бывает необходимым

RickHammerson commented on src/main/java/faang/school/paymentservice/dto/PaymentRequestDto.java:
```diff
@@ -0,0 +1,29 @@
+    private Long debitAccountId;
+
+    @NotNull
+    private Long creditAccountId;
```
Id бы ещё валидировать на @PozitiveOrZero

RickHammerson commented on src/main/java/faang/school/paymentservice/dto/event/EventDto.java:
```diff
@@ -0,0 +1,4 @@
+package faang.school.paymentservice.dto.event;
+
+public abstract class EventDto {
```
Идея хорошая, но класс без полей выглядит странно
Хотя бы туда записать время события и юзера, который это инициировал

RickHammerson commented on src/main/java/faang/school/paymentservice/model/PaymentRequest.java:
```diff
@@ -0,0 +1,72 @@
+    @Column(name = "updated_at")
+    private LocalDateTime updatedAt;
+
+    public boolean isClearScheduledAtNow() {
```
Метод называется "Шедулирован сейчас?", но сравнивает если дата не позже, чем сейчас
Как-то поиграть с названием

RickHammerson replied:
По типу "shouldClearNow"

RickHammerson commented on src/main/java/faang/school/paymentservice/redis/PaymentEventPublisher.java:
```diff
@@ -0,0 +1,17 @@
+public class PaymentEventPublisher extends AbstractEventPublisher {
+    public PaymentEventPublisher(ObjectMapper objectMapper,
+                                 RedisTemplate<String, Object> redisTemplate,
+                                 @Qualifier("paymentTopic") ChannelTopic topic) {
```
Перенос аргументов на новые строчки. Писал в другом PRе

RickHammerson commented on src/main/java/faang/school/paymentservice/service/PaymentService.java:
```diff
@@ -0,0 +1,127 @@
+    private final PaymentRequestMapper mapper;
+    private final PaymentEventPublisher eventPublisher;
+    private final PaymentServiceValidator verifier;
+
```
пробел лишний, либо я сноб :D

NadirCianid replied:
мне кажется так луче читается, когда переменные чуть сильнее отделены от методов

RickHammerson reviewed (commented):
Код супер понятный пишешь, кайф смотреть

NadirCianid committed:
Merge branch 'phoenix-master-bc4' into feature-BJS2-11714

# Conflicts:
#	src/main/java/faang/school/paymentservice/PaymentApplication.java
#	src/main/java/faang/school/paymentservice/controller/GlobalExceptionHandler.java
#	src/main/java/faang/school/paymentservice/controller/PaymentController.java
#	src/main/java/faang/school/paymentservice/dto/PaymentRequestDto.java
#	src/main/java/faang/school/paymentservice/model/PaymentRequest.java
#	src/main/java/faang/school/paymentservice/repository/PaymentRequestJpaRepository.java
#	src/main/resources/application.yaml
#	src/main/resources/db/changelog/changeset/V001__payment-service_payment_operation.sql

NadirCianid committed:
post merge fix

NadirCianid committed:
post merge fix

NadirCianid committed:
Изменил реализацию:
- добавил тип операции
- изменил id операции на UUID
- реализовал отправку авторизационного сообщения через rest
- изменил статусы операций

Sabina-Vishnya commented on src/main/java/faang/school/paymentservice/config/context/UserContext.java:
```diff
@@ -0,0 +1,21 @@
+@Component
+public class UserContext {
+
+    private final ThreadLocal<Long> userIdHolder = new ThreadLocal<>();
```
использование ThreadLocal чаще всего считается bad practice в боевых проектах, много рисков вносит
id клиента можно хранить на уровне запроса или сессии, а не потока
самое первое, что приходит - это использовать RequestContextHolder, но советую просто прочитать про альтернативны тред локалу и быть в курсе, тут это норм

NadirCianid replied:
Я скопировал это из другого нашего сервиса. Это не мой код. Да это такое себе оправдание) 
Но с feign нужно серьезно разбираться, да

Sabina-Vishnya commented on src/main/java/faang/school/paymentservice/exception/GlobalExceptionHandler.java:
```diff
@@ -34,9 +43,9 @@ public ErrorResponse handleHttpMessageNotReadableException(HttpMessageNotReadabl
     }
 
-    @ExceptionHandler(RuntimeException.class)
+    /*@ExceptionHandler(RuntimeException.class)
```
а почему?

NadirCianid replied:
Забыл убрать. Моя вина

NadirCianid committed:
Вернул обработку RuntimeException и добавил напоминание, чтобы написать тесты
